app/services/account_role_grantor.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class AccountRoleGrantor
+
+  attr_reader :account, :by
+
+  def initialize(account, by:)
+    @account = account
+    @by = by
+  end
+
+  def grant(user, role)
+    account_user = nil
+
+    within_transaction do
+      # Remove the old owner if we're assigning a new one
+      remove_old_user if role == AccountUser::ACCOUNT_OWNER
+
+      account_user = find_or_build_user_role(user, role)
+      account_user.save!
+
+      # In order to allow changes once an account is closed (i.e. becomes invalid
+      # per the Validator) we cannot run the validations directly on the account.
+      account_user.errors.add(:base, "Must have an account owner") if account.missing_owner?
+
+      raise ActiveRecord::Rollback if account_user.errors.any?
```

Comment:
I think I've figured out how to clean this up some. I'm also realizing we should probably not use `find_or_initialize_by` at all because if we're switching from purchaser -> BA or vice versa, it ends up updating the AccountUser rather than soft deleting, so I'll be changing this behavior a bit.